AngularJS/src/org/angular2/entities/metadata/psi/Angular2MetadataDirectiveBase.java

```diff
@@ -72,6 +76,12 @@ public Angular2DirectiveSelector getSelector() {
     return getCachedProperties().second;
   }
 
+  @NotNull
+  @Override
+  public Collection<? extends Angular2DirectiveAttribute> getAttributes() {
+    return getCachedValueWithClassDependencies(this::getAttributes);
```

Comment:
You can use simpler `getCachedClassBasedValue`, because your results do not depend on any additional things, just metadata object and class object.

Reply:
`return getCachedClassBasedValue(this::buildAttributes)`